add blog about how to extend kubevela by cuelang (#771)
oam-dev/kubevela: docs/blog/zh/extend-kubevela-by-cuelang-in-20-mins.md

> # 20分钟学会使用 cuelang 快速扩展 KubeVela

Reviewer: ```suggestion
# 如何在 20 分钟内给你的 K8s PaaS 上线一个新功能
```

>2020年12月14日 19:33, by @wonderflow

> 上个月，[KubeVela 正式发布](https://kubevela.io/#/blog/zh/kubevela-the-extensible-app-platform-based-on-open-application-model-and-kubernetes)了，作为一款简单易用且高度可扩展的应用管理平台与核心引擎，扩展性正是 KubeVela 的核心能力。本文将重点介绍如何使用 KubeVela 做扩展。

Reviewer: ```suggestion
上个月，[KubeVela 正式发布](https://kubevela.io/#/blog/zh/kubevela-the-extensible-app-platform-based-on-open-application-model-and-kubernetes)了，作为一款简单易用且高度可扩展的应用管理平台与核心引擎，可以说是广大平台工程师用来构建自己的云原生 PaaS 的神兵利器。那么本文就以一个实际的例子，讲解一下如何在 20 分钟内，为你基于 KubeVela 的 PaaS “上线“一个新能力！
```

      	cpu?: string
      }
```

乍一看挺长的，好像很复杂，但是不要着急，我们拆开来慢慢介绍，其实学起来很简单。

# 纯 Definition 不含 extension 字段的扩展模式

```
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: webservice
  annotations:
    definition.oam.dev/description: "`Webservice` is a workload type to describe long-running, scalable, containerized services that have a stable network endpoint to receive external network traffic from customers.
    If workload type is skipped for any service defined in Appfile, it will be defaulted to `Web Service` type."
spec:
  definitionRef:
    name: deployments.apps
```

这一部分满打满算11行，其中有3行是在介绍 `webservice`的功能，5行是固定的格式。只有2行是有信息量的：

```
  definitionRef:
    name: deployments.apps
```

这两行的意思代表了这个Definition背后用的CRD名称是什么，其格式是 `<resources>.<api-group>`.了解 K8s 的同学应该知道 K8s 中比较常用的是通过 `api-group`, `version` 和 `kind` 定位资源，而 `kind` 在 K8s restful API 中对应的是 `resources`。以大家熟悉 `Deployment` 和 `ingress` 为例，它的对应关系如下：

| api-group | kind | version | resourcs | 
| -------- | -------- | -------- | -------- |
| apps     | Deployment     | v1     | deployments |
|  networking.k8s.io/v1| Ingress | v1 | ingresses | 

> > 这里补充一个小知识，为什么有了 kind 还要加个 resourcs 的概念呢？ 因为一个CRD 除了kind本身还有一些像 status，replica 这样的字段希望跟spec本身解耦开来在restful API 中单独更新，所以 resources 除了 kind 对应的那一个，还会有一些 额外的resource，如 Deployment 的 status 表示为 `deployments/status`。

Reviewer: typo: resources

所以相信聪明的你已经明白了不含 extension 的情况下，Definition应该怎么写了，最简单的就是根据 K8s 的资源组合方式拼接一下，只要填下面三个尖括号的空格就可以了。

```
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: <这里写名称>
spec:
  definitionRef:
    name: <这里写resources>.<这里写api-group>
```

针对运维特征注册（TraitDefinition）也是这样。

```
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: <这里写名称>
spec:
  definitionRef:
    name: <这里写resources>.<这里写api-group>
```

所以把 `Ingress` 写进去就是：

```
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name:  ingress
spec:
  definitionRef:
    name: ingresses.networking.k8s.io
```

所以到这里，相信你已经掌握了一个不含extensions的基本扩展模式，而剩下部分就是围绕 cuelang 的抽象模板。

> # cuelang 的基本语法

Reviewer: CUE 的基础语法，移出去，单独作为一个小文档。而且这个小文档要先讲怎么用，再讲模板结构，再讲语法。工作量不比这篇文章少，可以单独弄，不需要耦合，本文也不需要依赖它，本文要自包含。

```
template: {
	apiVersion: "apps/v1"
	kind:       "Deployment"
	spec: {
		selector: matchLabels: {
			"app.oam.dev/component": parameter.name
		}
		template: {
			metadata: labels: {
				"app.oam.dev/component": parameter.name
			}
			spec: {
				containers: [{
					name:  parameter.name
					image: parameter.image
					if context["config"] != _|_ {
						env: parameter.config
					}
				}]
			}}}
}
```

4. 熟悉 K8s 的朋友可能已经发现，上面模板有点像 K8s 的Deployment。实际上这个就是一个 K8s Deployment 的模板。而模板中可替换的部分就是在 parameter 中的内容。让我们来为 parameter 赋值。
```
parameter:{
   name: "mytest"
   image: "nginx:v1"
   config:[{name:"a",value:"b"}]
}
```

5. **让我们依次把1~4步中的内容粘贴到一个cue文件中去，然后执行**：

```
$ cue export first.cue -e template --out yaml
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      containers:
      - name: mytest
        env:
        - name: a
          value: b
        image: nginx:v1
    metadata:
      labels:
        app.oam.dev/component: mytest
  selector:
    matchLabels:
      app.oam.dev/component: mytest
```

一个完整的 K8s deployment yaml 已经出现。

至此，我们已经学会了一些 cuelang 的基本用法并试着编写了一个Deployment模板，让我们回到正题，看下 KubeVela 中扩展该如何编写。

# KubeVela 的中使用 cuelang 模板扩展

> 在本文开头的 Definition 中定义的 extension 部分，其下层的 template 里面就是一个完整的 cue 文件内容。

Reviewer: 这一部分我的建议是

1. 一个讲解现有模板的例子，通过这个例子来解释模板的写法，让读者能够照猫画虎，而不是试图教会读者 CUE。具体应该采用”模板 vs 实际资源“对比的方式。这里有个例子：https://my.oschina.net/u/1787735/blog/4374952
2. 在（1）的基础上新增一个能力（要有趣，或者至少有一些实际意义，读者才有动力看），完整的走一下上述流程，结束本文。

4. `env` 可缺省的数组，数组的元素是个结构体，其中又嵌套了结构体。
5. `cpu` 可缺省的字符串。

所以在 Appfile 中，支持的也是上述这5个参数。

除此之外，KubeVela 还对 cuelang 的注释做了一些扩展，方便自动生成文档以及被 CLI 使用。 

## 模板部分

同样，KubeVela 固定的通过查找 Definition 中声明为 `output` 的变量来获得当前这个扩展能力的模板。

```
  output: {
      	apiVersion: "apps/v1"
      	kind:       "Deployment"
      	spec: {
      		selector: matchLabels: {
      			"app.oam.dev/component": context.name
      		}
      		template: {
      			metadata: labels: {
      				"app.oam.dev/component": context.name
      			}
      			spec: {
      				containers: [{
      					name:  context.name
      					image: parameter.image
      					if parameter["cmd"] != _|_ {
      						command: parameter.cmd
      					}
      					if parameter["env"] != _|_ {
      						env: parameter.env
      					}
      					if context["config"] != _|_ {
      						env: context.config
      					}
      					ports: [{
      						containerPort: parameter.port
      					}]
      					if parameter["cpu"] != _|_ {
      						resources: {
      							limits:
      								cpu: parameter.cpu
      							requests:
      								cpu: parameter.cpu
      						}}
      				}]
      		}}}
      }
```

> 这个模板并不复杂，值得一提的是，除了 parameter 以外，我们还看到了一个叫 context 的参数。`context` 是 KubeVela 内部自动生成的变量，里面存放了一些通用的环境上下文信息，如应用名称、秘钥等。

Reviewer: 要循序渐进，不要”值得一提“。先介绍简单模板，然后问一个问题 XXX 该怎么办？答案是 context。

Reviewer 2: https://github.com/oam-dev/kubevela/blob/master/pkg/appfile/service.go#L102     现在的context里应该只有组件名的， 应用名和密钥应用都还没放进去

> 至此，一个基本的 KubeVela 扩展能力的编写就介绍完了，其实也并不复杂，是不是？

Reviewer: 这句不要。换成能力上线之后，Appfile 咋写（同时也解释了 parameters ）

			}
			spec: {
				containers: [{
					name:  parameter.name
					image: parameter.image
				}]
			}}}
}
```

如上面的这个例子所示，KubeVela 中的模板参数就是通过 `parameter` 这个部分来完成的，而`parameter` 本质上就是作为引用，替换掉了 `output` 中的某些字段。

## 完整的 Definition 以及在 Appfile 试用

事实上，经过上面两部分的组合，我们已经可以写出一个完整的 Definition 文件：

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: mydeploy
spec:
  definitionRef:
    name: deployments.apps
  extension:
    template: |
        parameter: {
            name: string
            image: string
        }
        output: {
            apiVersion: "apps/v1"
            kind:       "Deployment"
            spec: {
                selector: matchLabels: {
                    "app.oam.dev/component": parameter.name
                }
                template: {
                    metadata: labels: {
                        "app.oam.dev/component": parameter.name
                    }
                    spec: {
                        containers: [{
                            name:  parameter.name
                            image: parameter.image
                        }]
                    }}}
        }
```

把这个 yaml 文件 apply 到 K8s 集群中。

> ```shell script

Author: good idea, I added the shell script.

		containers: [{
			name:  parameter.name
			image: parameter.image
			env: [
				for _, v in parameter.env {
					name:  v.name
					value: v.value
				},
			]
		}]
	}
}
```

Appfile 中的写法：
```
name: my-extend-app
services:
  mysvc:
    type: mydeploy
    name:  "mysvc"
    image: "nginx"
    env:
    - name: env1
      value: value1
    - name: env2
      value: value2
```

## KubeVela 内置的 `context` 变量

大家可能也注意到了，我们在 parameter 中定义的 name 每次在 Appfile中 实际上写了两次，一次是在 services 下面（每个service都以名称区分），
另一次则是在具体的`name`参数里面。事实上这里重复的不应该由用户再写一遍，所以 KubeVela 中还定义了一个内置的 `context`，里面存放了一些通用的环境上下文信息，如应用名称、秘钥等。
直接在模板中使用 context 就不需要额外增加一个 `name` 参数了， KubeVela 在运行渲染模板的过程中会自动传入。

```shell script
parameter: {
	image: string
}
output: {
  ...
	spec: {
		containers: [{
			name:  context.name
			image: parameter.image
		}]
	}
  ...
}
```

## KubeVela 中的注释增强

KubeVela 还对 cuelang 的注释做了一些扩展，方便自动生成文档以及被 CLI 使用。

```
 parameter: {
      	// +usage=Which image would you like to use for your service
      	// +short=i
      	image: string
      
      	// +usage=Commands to run in the container
      	cmd?: [...string]
       ...
      }
```

其中，`+usgae` 开头的注释会变成参数的说明，`+short` 开头的注释后面则是在 CLI 中使用的缩写。

# 总结

> 

Reviewer: ```suggestion
本文通过实际的案例和详细的讲述，为你介绍了在 KubeVela 中新增一个能力的详细过程与原理，以及能力模板的编写方法。

这里你可能还有个疑问，平台管理员这样添加了一个新能力后，平台的用户又该怎么能知道这个能力怎么使用呢？其实，在 KubeVela 中，它不仅能方便的添加新能力，**它还能自动为“能力”生成 Markdown 格式的使用文档！** 不信，你可以看下 KubeVela 本身的官方网站，所有在 `References/Capabilities`目录下能力使用说明文档（比如[这个](https://kubevela.io/#/en/developers/references/workload-types/webservice)），全都是根据每个能力的模板自动生成的哦。
```

    If workload type is skipped for any service defined in Appfile, it will be defaulted to `Web Service` type."
spec:
  definitionRef:
    name: deployments.apps
```

这一部分满打满算11行，其中有3行是在介绍 `webservice`的功能，5行是固定的格式。只有2行是有特定信息：

```yaml
  definitionRef:
    name: deployments.apps
```

这两行的意思代表了这个Definition背后用的CRD名称是什么，其格式是 `<resources>.<api-group>`.了解 K8s 的同学应该知道 K8s 中比较常用的是通过 `api-group`, `version` 和 `kind` 定位资源，而 `kind` 在 K8s restful API 中对应的是 `resources`。以大家熟悉 `Deployment` 和 `ingress` 为例，它的对应关系如下：

| api-group | kind | version | resources | 
| -------- | -------- | -------- | -------- |
| apps     | Deployment     | v1     | deployments |
|  networking.k8s.io | Ingress | v1 | ingresses | 

> 这里补充一个小知识，为什么有了 kind 还要加个 resources 的概念呢？
> 因为一个 CRD 除了 kind 本身还有一些像 status，replica 这样的字段希望跟 spec 本身解耦开来在 restful API 中单独更新，
> 所以 resources 除了 kind 对应的那一个，还会有一些额外的 resources，如 Deployment 的 status 表示为 `deployments/status`。

所以相信聪明的你已经明白了不含 extension 的情况下，Definition应该怎么写了，最简单的就是根据 K8s 的资源组合方式拼接一下，只要填下面三个尖括号的空格就可以了。

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: <这里写名称>
spec:
  definitionRef:
    name: <这里写resources>.<这里写api-group>
```

> 针对运维特征注册（TraitDefinition）也是这样。

Reviewer 3: `appliesToWorkloads`, 和 `workloadRefPath` 也挺重要，建议一并介绍一下。

apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name:  ingress
spec:
  definitionRef:
    name: ingresses.networking.k8s.io
```

所以到这里，相信你已经掌握了一个不含 extensions 的基本扩展模式，而剩下部分就是围绕 [CUE](https://cuelang.org/) 的抽象模板。

# 供 Appfile 使用的扩展模板（CUE Template）部分 

对 CUE 本身有兴趣的同学可以参考这篇[CUE 基础入门](https://wonderflow.info/posts/2020-12-15-cuelang-template/) 多做一些了解，限于篇幅本文对 CUE 本身不详细展开。

大家知道 KubeVela 的 Appfile 写起来很简洁，但是 K8s 的对象是一个相对比较复杂的 YAML，而为了保持简洁的同时又不失可扩展性，KubeVela 提供了一个从复杂到简洁的桥梁。
这就是 Definition 中 CUE Template 的作用。

## CUE 格式模板

让我们先来看一个 Deployment 的 YAML 文件，如下所示，其中很多内容都是固定的框架（模板部分），真正需要用户填的内容其实就少量的几个字段（参数部分）。

```yaml
apiVersion: apps/v1
kind: Deployment
meadata:
  name: mytest
spec:
  template:
    spec:
      containers:
      - name: mytest
        env:
        - name: a
          value: b
        image: nginx:v1
    metadata:
      labels:
        app.oam.dev/component: mytest
  selector:
    matchLabels:
      app.oam.dev/component: mytest
```

在 KubeVela 中，Definition 文件的固定格式就是分为 `output` 和 `parameter` 两部分。其中`output`中的内容就是“模板部分”，而 `parameter` 就是参数部分。

那我们来把上面的 Deployment YAML 改写成 Definition 中模板的格式。

```cue
output: {
	apiVersion: "apps/v1"
	kind:       "Deployment"

>     metadata: name: "mytest"

Reviewer 3: 可能要加几个空格：）

Author: thanks

	spec: {
		selector: matchLabels: {
			"app.oam.dev/component": "mytest"
		}
		template: {
			metadata: labels: {
				"app.oam.dev/component": "mytest"
			}
			spec: {
				containers: [{
					name:  "mytest"
					image: "nginx:v1"
					env: [{name:"a",value:"b"}]
				}]
			}}}
}
```

这个格式跟 json 很像，事实上这个是 CUE 的格式，而 CUE 本身就是一个 json 的超集。也就是说，CUE的格式在满足 JSON 规则的基础上，增加了一些简便规则，
使其更易读易用：

* C 语言的注释风格。
* 表示字段名称的双引号在没有特殊符号的情况下可以缺省。
* 字段值结尾的逗号可以缺省，在字段最后的逗号写了也不会出错。
* 最外层的大括号可以省略。

## CUE 格式的模板参数--变量引用

编写好了模板部分，让我们来构建参数部分，而这个参数其实就是变量的引用。

```
parameter: {
	name: string
	image: string
}
output: {
	apiVersion: "apps/v1"
	kind:       "Deployment"
	spec: {
		selector: matchLabels: {
			"app.oam.dev/component": parameter.name
		}
		template: {
			metadata: labels: {
				"app.oam.dev/component": parameter.name
			}
			spec: {
				containers: [{
					name:  parameter.name
					image: parameter.image
				}]
			}}}
}
```

> 如上面的这个例子所示，KubeVela 中的模板参数就是通过 `parameter` 这个部分来完成的，而`parameter` 本质上就是作为引用，替换掉了 `output` 中的某些字段。

Reviewer 3: 建议写一下通过 cue cli 的验证方法，直接合并起来再出错了，排查就比较费劲。
这个过程又很可能涉及到 context 引入，建议介绍一下 context。
提出一个问题：引入后，再调用 cue eval/export 验证又有点费劲，因为 context 不在当前文件，会报错。

Reviewer 3: 哈哈，后面看到你介绍 context 了，赞。不过，最后的问题还是存在。

Author: good idea, thanks. Added

NAME    	DESCRIPTION
mydeploy	description not defined
```

在 Appfile 中使用方式如下：

```yaml
name: my-extend-app
services:
  mysvc:
    type: mydeploy
    image: crccheck/hello-world
    name: mysvc
```

执行 `vela up` 就能把这个运行起来了：

```shell script
$ vela up -f docs/examples/blog-extension/my-extend-app.yaml
Parsing vela appfile ...
Loading templates ...

Rendering configs for service (mysvc)...
Writing deploy config to (.vela/deploy.yaml)

Applying deploy configs ...
Checking if app has been deployed...
App has not been deployed, creating a new deployment...
✅ App has been deployed 🚀🚀🚀
    Port forward: vela port-forward my-extend-app
             SSH: vela exec my-extend-app
         Logging: vela logs my-extend-app
      App status: vela status my-extend-app
  Service status: vela status my-extend-app --svc mysvc
```

我们来查看一下应用的状态，已经正常运行起来了（`HEALTHY Ready: 1/1`）：

```shell script
$ vela status my-extend-app
About:

  Name:      	my-extend-app
  Namespace: 	env-application
  Created at:	2020-12-15 16:32:25.08233 +0800 CST
  Updated at:	2020-12-15 16:32:25.08233 +0800 CST

Services:

  - Name: mysvc
    Type: mydeploy
    HEALTHY Ready: 1/1
```

> # Definition 模板中的高级用法

Reviewer 3: 这一部分很赞～
